src/NServiceBus.Core.Analyzer/AwaitOrCaptureTasksAnalyzer.cs

```diff
@@ -0,0 +1,90 @@
+namespace NServiceBus.Core.Analyzer
+{
+    using System;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.Diagnostics;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class AwaitOrCaptureTasksAnalyzer : DiagnosticAnalyzer
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(diagnostic);
+
+        public override void Initialize(AnalysisContext context) => context.RegisterSyntaxNodeAction(Analyze, SyntaxKind.InvocationExpression);
+
+        void Analyze(SyntaxNodeAnalysisContext context)
+        {
+            if (!(context.Node is InvocationExpressionSyntax call))
+            {
+                return;
+            }
+
+            if (!(call.Parent is ExpressionStatementSyntax))
+            {
+                return;
+            }
+
+            foreach (var token in call.Expression?.DescendantTokens() ?? Enumerable.Empty<SyntaxToken>())
+            {
+                // check syntax tree (cheap) first for possible call requiring await and then check semantic model (expensive) to confirm
+                if (CouldBeMethodRequiringAwait(token) && IsMethodRequiringAwait(call, context))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(diagnostic, call.GetLocation(), call.ToString()));
+                    return;
+                }
+            }
+        }
+
+        static bool CouldBeMethodRequiringAwait(SyntaxToken syntaxToken) =>
+            syntaxToken.Kind() == SyntaxKind.IdentifierToken && methodNames.Contains(syntaxToken.Text);
+
+        static bool IsMethodRequiringAwait(ExpressionSyntax call, SyntaxNodeAnalysisContext context) =>
+            context.SemanticModel.GetSymbolInfo(call).Symbol is IMethodSymbol methodSymbol &&
+                methods.Contains(methodSymbol.GetFullName());
+
+        static readonly DiagnosticDescriptor diagnostic = new DiagnosticDescriptor(
+            "NSB0001",
+            "Await or capture tasks",
+            "A Task returned by an NServiceBus method is not awaited or assigned to a variable.",
+            "NServiceBus.Code",
+            DiagnosticSeverity.Error,
+            true,
+            "Tasks returned by specific NServiceBus methods must be awaited. Failure to await these Tasks may result in message loss due to swallowed exceptions. If the task is assigned to a variable, this diagnostic is not shown, but the Task must still be awaited.");
```

Comment:
Can we drop "specific" from this sentence?

Reply:
"may result in message loss due to swallowed exceptions" 

How accurate is this? AFAIK swallowed exceptions aren't the only reason not awaiting a `Task` could cause message loss. There's also the race between the Task being completed and the pipeline being completed after the handler returns.

Maybe we can just shorten it to "Failure to await these Tasks may result in message loss."

Reply:
> Can we drop "specific" from this sentence?

Are we sure there are, and never will be, any NSB methods which are OK to fire and forget? If we are, then sure, let's drop it.

> Maybe we can just shorten it to "Failure to await these Tasks may result in message loss."

👍 to make that change.

Reply:
👍 to also remove the "specific"